### DISCUSSION TITLE
How could I smooth the 3D image using the pysph?

### DISCUSSION THREAD

**Discussion by KirsCAI** [pypr/pysph#385] [General]
Hi,
I'm using the IllustrisTNG hydrodynamic simulation. And the particle of the simulation is not smoothed. The parameter of the simulation are the xyz coordinate, mass, h parameter(which is calculated from the star particle enclosing the 64±1 nearest dark matter particles.) and density(The local total comoving mass density, estimated using the standard cubic-spline SPH kernel over all particles/cells within a radius of h.). 

First question is that if I do not want to do the fluid simulated calculation(like gradient calculation) and just do the 3D image smoothing of kernel(like cubic-spline kernel), what package of pySPH should I use? 

Second question is that what the meaning of density which is calculated from cubic-spline kernel. Is not the calculated result of kernel a distribution？ I feel so confused.

Third question is that if I want to calculate the h parameter and distribution by myself, what pysph package should I use?

Thank you very much.

**Comment by pawansnegi** [comment]
Hello,

The pysph software can be in general used for any numerical method that
require neighbour particle information to construct shape function.
However, it has a very extensive support for SPH method.

Please find the answer to your questions within

On Sun, 5 Nov, 2023, 7:07 pm KirsCAI, ***@***.***> wrote:

> Hi,
> I'm using the IllustrisTNG hydrodynamic simulation. And the particle of
> the simulation is not smoothed.
>

Do you want a uniform particle distribution on the existing field or want
the field smoothed out on existing particles? Method for both of them are
similar.

> The parameter of the simulation have the xyz coordinate, mass, h
> parameter(which is calculated from the star particle enclosing the 64±1
> nearest dark matter particles.) and density(The local total comoving mass
> density, estimated using the standard cubic-spline SPH kernel over all
> particles/cells within a radius of h.).
>
> First question is that if I do not want to do the fluid simulated
> calculation(like gradient calculation) and just do the 3D image smoothing
> of kernel(like cubic-spline kernel), what package of pySPH should I use?
>

You can use SPHEvaluator in pysph.tools.spseveluator. It can compute any
set of given equations and overwrite the output property.

> Second question is that what the meaning of density which is calculated
> from cubic-spline kernel. Is not the calculated result of kernel a
> distribution？ I feel so confused.
>
Density is the kernel density.

Third question is that if I want to calculate the h parameter and
> distribution by myself, what pysph package should I use?
>

You can set h directly to the particles array as a property before passing
it to the SPHEvaluator.
